src/cp_als.jl

```diff
@@ -0,0 +1,80 @@
+export cp_als
+
+## The Alternating Least-Squares (ALS) algorithm for CP-decomposition with ITensors.jl
+# Based on "Tensor Decompositions and Applications", Tamara G. Kolda and Brett W. Bader. SIAM Review 51(3), 2009
+# Porting the code from https://www.kaggle.com/nicw102168/rank-of-random-2x2x2-tensors
+
+using ITensors
+using ITensors.LinearAlgebra
+
+
+function cp_als(X, R; maxiter=1000, test_period=1, tol=1e-10)
+    # initialize with spheric-random colunms
+    A = map([randomITensor(Iₙ, Index(R)) for Iₙ in inds(X)]) do An
+        aλ = columnnorms(An)
+        rr = Index(R)
+        iλ = diagITensor(aλ.^-1, rr, inds(An)[2])
+        replaceind(An, inds(An)[2], rr) * iλ
+    end
+    cp_als_(X, R, A, maxiter=maxiter, test_period=test_period, tol=tol)
+end
+
+function cp_als_(X, R, A; maxiter=1000, test_period=1, tol=1e-10)
+    N = order(X)
+
+    rind = [inds(An)[2] for An in A]
+    λ = nothing
+    iteration = nothing
+
+    for it in 1:maxiter
+        iteration=it
+
+        for n in 1:N
+            V = reduce((a,b)->a.*b, array(A[m])' * array(A[m]) for m in 1:N if m != n) ## is there a better, ITensor way to do that?
```

Comment:
Could you explain the equation you are trying to compute here?

Reply:
I tried to closely follow the algorithm presented in Kolda&Bader, Figure 3.3 here https://www.kolda.net/publication/TensorReview.pdf

A(n) are matrices of InxR, where R is the order of the approximation. This loop goes through each dimension, and V is computed as the Hadamard product of the the matrix products of every other dimension, skipping the current `n`.

Therefore `[array(A[m])' * array(A[m]) for m in 1:N if m != n]` is this collection of the matrix products by themselves, skipping the current dimension. We take that, and compute the total Hadamard product. We could probably define a `hadamard_collection(xx) = reduce(.*, xx)` or something like that.

V is an array of